rs/bitcoin/adapter/src/header_cache.rs

```diff
@@ -450,83 +382,148 @@ impl<Header: BlockchainHeader> LMDBHeaderCache<Header> {
     }
 }
 
-impl<Header: BlockchainHeader + Send + Sync> HeaderCache for LMDBHeaderCache<Header> {
-    type Header = Header;
+pub struct HybridHeaderCache<Header> {
+    in_memory: RwLock<InMemoryHeaderCache<Header>>,
+    on_disk: Option<LMDBHeaderCache>,
+    genesis_hash: BlockHash,
+}
 
-    fn get_genesis(&self) -> PureHeader {
-        self.genesis.clone().into_pure_header()
+impl<Header: BlockchainHeader> HybridHeaderCache<Header> {
+    pub fn new(genesis: Header, cache_dir: Option<PathBuf>, log: ReplicaLogger) -> Self {
+        let genesis_hash = genesis.block_hash();
+        let on_disk = cache_dir.map(|dir| LMDBHeaderCache::new(dir, log));
+        // Try reading the anchor (tip of the chain) from disk.
+        // If it doesn't exist, use genesis header.
+        let anchor = on_disk
+            .as_ref()
+            .and_then(|cache| {
+                log_err_except!(
+                    cache.run_ro_txn(|tx| cache.tx_get_tip(tx)),
+                    cache.log,
+                    LMDBCacheError::Lmdb(lmdb::Error::NotFound),
+                    "tx_get_tip()"
+                )
+            })
+            .unwrap_or_else(|| HeaderData {
+                header: genesis.clone(),
+                height: 0,
+                work: genesis.work(),
+            });
+        let in_memory = RwLock::new(InMemoryHeaderCache::new_with_anchor(anchor));
+        Self {
+            in_memory,
+            on_disk,
+            genesis_hash,
+        }
     }
+}
 
-    fn get_header(&self, hash: BlockHash) -> Option<HeaderNode<Header>> {
-        log_err_except!(
-            self.run_ro_txn(|tx| self.tx_get_header(tx, hash)),
-            self.log,
-            LMDBCacheError::Lmdb(lmdb::Error::NotFound),
-            format!("tx_get_header({hash})")
-        )
+impl<Header: BlockchainHeader + Send + Sync + 'static> HybridHeaderCache<Header> {
+    /// Get the genesis header.
+    pub fn get_genesis(&self) -> PureHeader {
+        // The unwrap below is safe because:
+        // - if the on-disk cache is configured and has persisted data, it should
+        //   contain the genesis header;
+        // - otherwise the in-memory cache hasn't been pruned, and should contain
```

Comment:
Even if the in-memory cache has been pruned, it should contain the chain leading all the way back to the genesis block header, right?